circuit/environment/src/traits/to_bits.rs

```
@@ -59,6 +59,36 @@ pub trait ToBitsRaw: ToBits {
     fn write_bits_raw_be(&self, vec: &mut Vec<Self::Boolean>);
 }
 
+pub trait ToBitArray: ToBits {
+    /// Returns `self` as a bit array in little-endian order, resizing to the given length.
+    fn to_bit_array_le(&self, length: u32) -> Vec<Self::Boolean> {
+        use crate::traits::Inject;
+        // Get the bits.
+        let mut bits = self.to_bits_le();
+        // Resize to the given length.
+        bits.resize(length as usize, Self::Boolean::constant(false));
```

I think the padding with const is fine, as array sizes are fixed at deployment time.

I'm worried that **trimming** the bits might be unintended behaviour. Can you add an example program using the new serialization in test_v11.rs ? A developer can't easily see how many bits their types serialize to, so we could consider adding an ensure!(bits.len() <= length);

The same comment applies to other places where you resize.

---

Fair point! I think it all boils down to the UX we want to give developers.
Do developers need to know the exact size of the data types they want to serialize or deserialize into?
When you introduce arrays and composite types, this can be fairly tricky to do manually.
But a compiler should be able to do this for you.
The trimming can be helpful if the data type is too large to fit into our max array size.
I don't have a strong preference, I'll leave to reviewers to arbitrate.

Is there a specific behavior you wanted to test? Like serializing into smaller bits?
There are [these](https://github.com/ProvableHQ/snarkVM/blob/feat/bytes/synthesizer/process/src/tests/test_serializers.rs) tests that check for serde equivalence

---

I'm specifically worried about a case where both in the client/SDK and server/deployment, a developer aims to serialize a particular message to sign and e.g. some serialized struct fields are silently ignored because of the trimming. The only way out I see at the moment is to not do the trimming... 

If it is functionality we care about in the future, we could add yet another opcode or variant which does trim... However, the developer can always manipulate and test their input data to trim it appropriately, therefore intentionally skipping certain fields. And we can build utilities on the client/SDK side to make that easier.

> There are [these](https://github.com/ProvableHQ/snarkVM/blob/feat/bytes/synthesizer/process/src/tests/test_serializers.rs) tests that check for serde equivalence

Thx that seems to suffice! Do I understand correctly that the `evaluate_function` invocation tests the serde path?

---

I see your point. I'm getting rid of these traits and impls in the `utilities` and `circuit` crates. 

**How do you feel about the opcodes `serialize.bits*` and `deserialize.bits*` doing trimming?**

The program that is run in `evaluate_function` and `execute_function` tests that a given input type can be serialized and deserialized equivalently in the AVM.

---

For context,
```
serialize r0 (field) into r1 ([boolean; u64]);
````
Should it error or give you the lower 64 bits of the field? (edited) 

```
derialize r0 ([boolean; 4]) into r1 (field);
```
Should it error or pad appropriately to the right number of field bits?

```
deserialize r0 ([boolean; 512]) into r1 (field);
```
Should it error or use the lower 253 bits and construct a field element? (edited)

---

The `native.*` variants of the `keccak` and `sha3` hashes are implemented with trimming and extension. Should they be restricted to the exact number of bits?

---

From an offline discussion with @vicsn 
```
IMO for the first implementation the middle one should pad, first and third should error.
Because trimming can still be done manually at the application level for usecases people care about. But then its at least safe by default.
Similar to how someone has to explicitly use .raw to start making exceptions
```

---

I opted for requiring the correct byte sizes for all opcodes for consistency.
This is implemented in this PR: https://github.com/ProvableHQ/snarkVM/pull/2939